### ISSUE TITLE
Exception when too many files requested to AtomCache

### ISSUE THREAD

**Issue by lafita** [biojava/biojava#395]
When the same **AtomCache** is used to request for many structures (not at the same time) an Exception is thrown and the following message is given: **Too many open files**. 

It could indicate a resource or memory leak in **AtomCache**, by maintaining some files open (although the structure has already been downloaded). I get those Exceptions when downloading, in parallel 4 at the same time, more than 2K structures. I do not remember the Exception when I calculated the census the first time (6 months ago), although I am not sure if I was using exactly the same code. Any ideas?

This is the **stacktrace** of the Exception:

```java
java.io.FileNotFoundException: /scratch/cache/data/structures/divided/mmCIF/rl/1rlr.cif.gz (Too many open files)
	at java.io.FileInputStream.open(Native Method) ~[?:1.6.0_38]
	at java.io.FileInputStream.<init>(FileInputStream.java:140) ~[?:1.6.0_38]
	at org.biojava.nbio.core.util.FlatFileCache.addToCache(FlatFileCache.java:63) [classes/:?]
	at org.biojava.nbio.core.util.InputStreamProvider.getInputStreamFromFile(InputStreamProvider.java:247) [classes/:?]
	at org.biojava.nbio.core.util.InputStreamProvider.getInputStream(InputStreamProvider.java:163) [classes/:?]
	at org.biojava.nbio.structure.io.LocalPDBDirectory.getInputStream(LocalPDBDirectory.java:377) [classes/:?]
	at org.biojava.nbio.structure.io.LocalPDBDirectory.getStructureById(LocalPDBDirectory.java:343) [classes/:?]
	at org.biojava.nbio.structure.align.util.AtomCache.loadStructureFromCifByPdbId(AtomCache.java:1038) [classes/:?]
	at org.biojava.nbio.structure.align.util.AtomCache.getStructureForPdbId(AtomCache.java:1019) [classes/:?]
	at org.biojava.nbio.structure.scop.ScopDomain.loadStructure(ScopDomain.java:245) [classes/:?]
	at org.biojava.nbio.structure.align.client.StructureName.loadStructure(StructureName.java:349) [classes/:?]
	at org.biojava.nbio.structure.align.util.AtomCache.getStructure(AtomCache.java:394) [classes/:?]
	at demo.CeSymmMain$CeSymmWorker.run(CeSymmMain.java:1061) [classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471) [?:1.6.0_38]
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334) [?:1.6.0_38]
	at java.util.concurrent.FutureTask.run(FutureTask.java:166) [?:1.6.0_38]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1146) [?:1.6.0_38]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [?:1.6.0_38]
	at java.lang.Thread.run(Thread.java:701) [?:1.6.0_38]
``

**Comment by lafita** [comment]
I ran **TestAtomCachePerformance** (although it fetches only 100 structures at maximum) and everything is working fine.

**Comment by lafita** [comment]
This Exception is also thrown some of the times, in the same case:

```java
Caused by: java.net.SocketException: Too many open files
	at java.net.Socket.createImpl(Socket.java:414) ~[?:1.6.0_38]
	at java.net.Socket.connect(Socket.java:544) ~[?:1.6.0_38]
	at sun.net.NetworkClient.doConnect(NetworkClient.java:173) ~[?:1.6.0_38]
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:427) ~[?:1.6.0_38]
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:529) ~[?:1.6.0_38]
	at sun.net.www.http.HttpClient.<init>(HttpClient.java:213) ~[?:1.6.0_38]
	at sun.net.www.http.HttpClient.New(HttpClient.java:306) ~[?:1.6.0_38]
	at sun.net.www.http.HttpClient.New(HttpClient.java:325) ~[?:1.6.0_38]
	at sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:956) ~[?:1.6.0_38]
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:892) ~[?:1.6.0_38]
	at sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:810) ~[?:1.6.0_38]
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1227) ~[?:1.6.0_38]
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:397) ~[?:1.6.0_38]
	at org.biojava.nbio.structure.align.util.HTTPConnectionTools.getInputStream(HTTPConnectionTools.java:186) ~[classes/:?]
	at org.biojava.nbio.structure.align.util.HTTPConnectionTools.getInputStream(HTTPConnectionTools.java:158) ~[classes/:?]
	at org.biojava.nbio.structure.scop.RemoteScopInstallation.requestRemoteDomainByScopID(RemoteScopInstallation.java:208) ~[classes/:?]
	at org.biojava.nbio.structure.scop.RemoteScopInstallation.getDomainByScopID(RemoteScopInstallation.java:220) ~[classes/:?]
```

**Comment by josemduarte** [comment]
The first stack trace you post is for a file not found, that doesn't look like a problem with too many open files. The second one does.

This reminds me of an issue fixed recently, where we had a resource leak with `DownloadChemCompProvider` (see commit d3bb20e2369).  I suppose you are running on latest HEAD, aren't you?

**Comment by lafita** [comment]
Yes, I am running on the latest HEAD. 

If you look at the first line of the first stack trace there is a strange line printed that says: _Too many files open_. It cannot be a regular `FileNotFoundException`, since running the code on that entry alone does not reproduce the Exception. 

Furthermore, instead of using a global `AtomCache `for all entries I used a different `AtomCache` instance for each entry and the Exceptions do not occur. While it fixes this particular error, I will try to debug the resource leak and see where it could be (although it is difficult). Maybe that other leak in `DownloadChemCompProvider `helps. Thanks @josemduarte

**Comment by lafita** [comment]
As another hint, when I run the code with **Debug** mode, the Exceptions are not thrown anymore.

**Comment by lafita** [comment]
I ran overnight 5K structures and got no Exception like the ones described, so it seems hard to reproduce.

**Comment by josemduarte** [comment]
Did you try running with many threads concurrently? That might trigger the error faster.

**Comment by lafita** [comment]
Yes, I was running 4 threads concurrently when the Exception was thrown. Now I tried increasing the threads to 10 and 20, but I cannot reproduce the error anymore. It might be some error with my OS in that particular day, and after restarting the session it was solved. So hopefully there is no resource leak in AtomCache. 

I would consider closing the issue, but having it in mind in case some similar Exception occurs in the future. @abradle I remember you were parsing PDB structures in a script not long ago, did you notice any similar Exception?

**Comment by andreasprlic** [comment]
There was a file leak with the  DownloadChemCompProvider, as Jose has pointed out above, but it has been recently fixed (d3bb20e2369f0c9958f729d1f043ef529c632d9d). If we can't reproduce the "too many files open" any longer, I'd blame it on that and close this one.

**Comment by sbliven** [comment]
I've found a more likely culprit than the leak: ReducedChemCompProvider. I did some profiling using [file-leak-detector](http://file-leak-detector.kohsuke.org/) and I see an ever-increasing number of handles to the chemcomp cifs:

```
799 descriptors are open
#1 /home/spencer/git/biojava/biojava-structure/target/classes/chemcomp/HOH.cif.gz by thread:main on Tue Feb 09 12:45:12 CET 2016
	at java.io.FileInputStream.<init>(FileInputStream.java:147)
	at java.io.FileInputStream.<init>(FileInputStream.java:101)
	at sun.net.www.protocol.file.FileURLConnection.connect(FileURLConnection.java:90)
	at sun.net.www.protocol.file.FileURLConnection.getInputStream(FileURLConnection.java:188)
	at java.net.URLClassLoader.getResourceAsStream(URLClassLoader.java:233)
	at java.lang.Class.getResourceAsStream(Class.java:2111)
	at org.biojava.nbio.structure.io.mmcif.ReducedChemCompProvider.getChemComp(ReducedChemCompProvider.java:63)
	at org.biojava.nbio.structure.io.mmcif.ChemCompGroupFactory.getChemComp(ChemCompGroupFactory.java:58)
	at org.biojava.nbio.structure.io.ChargeAdder.addCharges(ChargeAdder.java:51)
	at org.biojava.nbio.structure.io.mmcif.SimpleMMcifConsumer.addCharges(SimpleMMcifConsumer.java:979)
	at org.biojava.nbio.structure.io.mmcif.SimpleMMcifConsumer.documentEnd(SimpleMMcifConsumer.java:733)
	at org.biojava.nbio.structure.io.mmcif.SimpleMMcifParser.triggerDocumentEnd(SimpleMMcifParser.java:1132)
	at org.biojava.nbio.structure.io.mmcif.SimpleMMcifParser.parse(SimpleMMcifParser.java:355)
	at org.biojava.nbio.structure.io.MMCIFFileReader.getStructure(MMCIFFileReader.java:111)
	at org.biojava.nbio.structure.io.LocalPDBDirectory.getStructureById(LocalPDBDirectory.java:345)
	at org.biojava.nbio.structure.align.util.AtomCache.loadStructureFromCifByPdbId(AtomCache.java:1038)
	at org.biojava.nbio.structure.align.util.AtomCache.getStructureForPdbId(AtomCache.java:1019)
	at org.biojava.nbio.structure.SubstructureIdentifier.loadStructure(SubstructureIdentifier.java:315)
	at org.biojava.nbio.structure.align.client.StructureName.loadStructure(StructureName.java:374)
	at org.biojava.nbio.structure.align.util.AtomCache.getStructure(AtomCache.java:394)
	at org.biojava.nbio.structure.align.util.AtomCache.getStructure(AtomCache.java:378)
	at org.biojava.nbio.structure.test.ecod.EcodParseTest.testVersion(EcodParseTest.java:87)
	at org.biojava.nbio.structure.test.ecod.EcodParseTest.main(EcodParseTest.java:69)
```
(Now at over 2000 a few minutes later.)

Each compound is open multiple times, especially the waters. There is definitely a resource leak here. It also points to this as a spot where some light caching could make a major improvement.
